File: Week 6/Python Codes/EquivalentSubarrays.py
def countDistinctSubarray(arr, n): 
    # Count distinct elements in whole array 
    arr_dict = dict() 
    for i in range(n): 
        arr_dict[arr[i]] = 1

    k = len(arr_dict) 
    subarr_dict = dict() 
    
    # Use sliding window concept to find count of subarrays having k distinct elements. 
    ans, right, window = 0, 0, 0
    
    for left in range(n): 
        while (right < n and window < k): 
            if arr[right] in subarr_dict: 
                subarr_dict[arr[right]] += 1
            else: 
                subarr_dict[arr[right]] = 1
    
            if (subarr_dict[arr[right]] == 1): 
                window += 1
    
            right += 1
            
        # If window size equals to array distinct element size, then update answer 
        if (window == k): 
            ans += (n - right + 1) 
    
        # Decrease the frequency of previous element for next sliding window 
        subarr_dict[arr[left]] -= 1
    
        # If frequency is zero then decrease the window size 
        if (subarr_dict[arr[left]] == 0): 
            window -= 1
    return ans 

File: Week 6/C++ Codes/EquivalentSubarrays.cpp
#include<bits/stdc++.h>
using namespace std;

int countDistinctSubarray(int arr[], int n) {
        // Count distinct elements in whole array 
        unordered_map<int, int>  vis; 
        for (int i = 0; i < n; ++i) 
            vis[arr[i]] = 1; 
        int k = vis.size(); 
      
        // Reset the container by removing all elements 
        vis.clear(); 
      
        // Use sliding window concept to find count of subarrays having k distinct elements. 
        int ans = 0, right = 0, window = 0; 
        for (int left = 0; left < n; ++left) { 
            while (right < n && window < k) { 
                vis[arr[right]]++; 
                
                if (vis[arr[right]] == 1) 
                    window++; 
                right++; 
            } 
      
            // If window size equals to array distinct element size, then update answer 
            if (window == k) 
                ans += (n - right + 1); 
      
            // Decrease the frequency of previous element for next sliding window 
            vis[arr[left]]--; 
      
            // If frequency is zero then decrease the window size 
            if (vis[arr[left]] == 0) 
                    window--; 
        } 
        return ans; 
    }